core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/RowDenormalizerAggregator.java

```diff
@@ -0,0 +1,133 @@
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.batch.BatchAggregator;
+import co.cask.cdap.etl.api.batch.BatchAggregatorContext;
+import co.cask.cdap.etl.api.batch.BatchRuntimeContext;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Batch RowDenormalizer Aggregator Plugin - It is used to de-normalize data based on the key column.
+ */
+@Plugin(type = BatchAggregator.PLUGIN_TYPE)
+@Name("RowDenormalizer")
+@Description("This plugin is used to convert raw data into denormalized data based on a key column. " +
+  "User is able to specify the list of fields that should be used in the denormalized record, with " +
+  "an option to use an alias for the output field name. " +
+  "For example, " +
+  "'ADDRESS' in the input is mapped to 'addr' in the output schema. The denormalized data is easier to query.")
+public class RowDenormalizerAggregator extends BatchAggregator<String, StructuredRecord, StructuredRecord> {
+
+  private final RowDenormalizerConfig conf;
+  private Map<String, String> outputMappings;
+  private Set<String> outputFields;
+  private Schema outputSchema;
+  private String keyField;
+
+  public RowDenormalizerAggregator(RowDenormalizerConfig conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void prepareRun(BatchAggregatorContext context) throws Exception {
+    if (conf.numPartitions != null) {
+      context.setNumPartitions(conf.numPartitions);
+    }
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    validateInputFields(pipelineConfigurer.getStageConfigurer().getInputSchema());
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(initializeOutputSchema());
+  }
+
+  @Override
+  public void initialize(BatchRuntimeContext context) throws Exception {
+    outputFields = conf.getOutputFields();
+    outputMappings = conf.getOutputFieldMappings();
+    keyField = conf.getKeyField();
+    outputSchema = initializeOutputSchema();
+  }
+
+  @Override
+  public void groupBy(StructuredRecord record, Emitter<String> emitter) throws Exception {
+    emitter.emit(record.get(keyField).toString());
+  }
+
+  @Override
+  public void aggregate(String groupKey, Iterator<StructuredRecord> iterator,
+                        Emitter<StructuredRecord> emitter) throws Exception {
+    if (!iterator.hasNext()) {
+      return;
+    }
+    StructuredRecord.Builder builder = StructuredRecord.builder(outputSchema);
+    builder.set(conf.getKeyField(), groupKey);
+    while (iterator.hasNext()) {
+      StructuredRecord record = iterator.next();
+      String fieldName = record.get(conf.getFieldName());
+      fieldName = outputMappings.containsKey(fieldName) ? outputMappings.get(fieldName) : fieldName;
+      if (outputFields.contains(fieldName)) {
+        builder.set(fieldName, record.get(conf.getFieldValue()));
```

Comment:
If every output field is supposed to be a string, need to check the type of this field and do the appropriate conversion to string.